File: base/src/main/vsdk/toolkit/environment/geometry/polyhedralBoundedSolid/TopologicalIntegrityStrategy.java
package vsdk.toolkit.environment.geometry.polyhedralBoundedSolid;

public class TopologicalIntegrityStrategy
    implements PolyhedralBoundedSolidValidationStrategy
{
    @Override
    public boolean validate(PolyhedralBoundedSolid solid,
        PolyhedralBoundedSolidNumericPolicy.ToleranceContext numericContext,
        StringBuilder msg)
    {
        PolyhedralBoundedSolidTopologicalValidator
            .remakeEmanatingHalfedgesReferences(solid);
        if ( !PolyhedralBoundedSolidTopologicalValidator
            .validateTopologicalIntegrity(solid) ) {
            msg.append("  - Topological integrity test failed.\n");
            return false;
        }
        return true;
    }
}

File: typescript/base/src/main/vsdk/toolkit/environment/geometry/volume/polyhedralBoundedSolid/_TopologicalIntegrityStrategy.ts
//= References:                                                             =
//= [MANT1988] Mantyla Martti. "An Introduction To Solid Modeling",         =
//=     Computer Science Press, 1988.                                       =

import type { StringBuilder } from "../../../../../../java/lang/StringBuilder.js";
import type { PolyhedralBoundedSolid } from "./PolyhedralBoundedSolid.js";
import type { ToleranceContext } from "./PolyhedralBoundedSolidNumericPolicy.js";
import type { _PolyhedralBoundedSolidValidationStrategy } from "./_PolyhedralBoundedSolidValidationStrategy.js";
import { _PolyhedralBoundedSolidTopologicalValidator } from "./_PolyhedralBoundedSolidTopologicalValidator.js";

/**
Applies topological consistency checks for the half-edge data structure of
[MANT1988].10.2.1 and [MANT1988].10.2.2.
*/
export class _TopologicalIntegrityStrategy implements _PolyhedralBoundedSolidValidationStrategy {
    /**
    Validates the basic incidence and cycle properties assumed by the half-edge
    representation in [MANT1988].10.2.1 and [MANT1988].10.2.2.
    */
    public validate(solid: PolyhedralBoundedSolid, _numericContext: ToleranceContext, msg: StringBuilder): boolean {
        _PolyhedralBoundedSolidTopologicalValidator.remakeEmanatingHalfedgesReferences(solid);
        if (!_PolyhedralBoundedSolidTopologicalValidator.validateTopologicalIntegrity(solid)) {
            msg.append("  - Topological integrity test failed.\n");
            return false;
        }
        return true;
    }
}
